src/transformers/models/bart/modeling_bart.py

```diff
@@ -489,19 +496,130 @@ def _upad_input(self, query_layer, key_layer, value_layer, attention_mask, query
         )
 
 
+class BartSDPAAttention(BartAttention):
+    def forward(
+        self,
+        hidden_states: torch.Tensor,
+        key_value_states: Optional[torch.Tensor] = None,
+        past_key_value: Optional[Tuple[torch.Tensor]] = None,
+        attention_mask: Optional[torch.Tensor] = None,
+        layer_head_mask: Optional[torch.Tensor] = None,
+        output_attentions: bool = False,
+        **kwargs,
+    ) -> Tuple[torch.Tensor, Optional[torch.Tensor], Optional[Tuple[torch.Tensor]]]:
+        """Input shape: Batch x Time x Channel"""
+        if output_attentions or layer_head_mask is not None:
+            return super().forward(
```

Comment:
Not in favor of silently switching the attention algorithm here, let's throw an error instead similarly to how it's done here: https://github.com/huggingface/transformers/blob/606d90845ff149ee08629edab95b576f788243a0/src/transformers/models/bart/modeling_bart.py#L311

We should probably also deprecate the "silent" disabling in Llama: https://github.com/huggingface/transformers/blob/606d90845ff149ee08629edab95b576f788243a0/src/transformers/models/llama/modeling_llama.py#L458 in a follow-up PR

Reply:
@patrickvonplaten SDPA attention would be enabled by default - throwing an error would be a breaking change for users using `head_mask` or `output_attentions`.

Reply:
Ah that makes sense I see! Hmm ok, I think that's a tricky one then. 

I'm def in favor of enabling SDPA by default, guess that then leaves us three options:
- a) leave as is, but add a warning that we're falling back to vanilla attention
- b) silently set `output_attentions=False` as done here: https://github.com/huggingface/transformers/blob/606d90845ff149ee08629edab95b576f788243a0/src/transformers/models/llama/modeling_llama.py#L458
- c) still raise an error if people use `output_attentions`

I think b) is a bad idea.

Between a) and c) it's a tough one IMO because I really don't think a) is the right choice long-term as it's black magic and hides the complexity. To me if you use Flash Attention / SDPA, you cannot get the attention weights which is a logical consequence of the method being used, I think it's better to explain this to users instead of hiding it. Also a) can lead to a couple hard-to-debug errors:
- You get OOM if you pass `output_attentions=True` but have no idea why
- You get slow-downs with `output_attentions=True` that are very significant
Also a) is inconsistent with how it's implemented for FA2

On the other hand going for a) would indeed by a big breaking change, but if we explain it well and give the user an easy way to change to "vanilla" attention on the fly it's the better choice. 

Overall, I'd advocate for:
- Let's add SDPA but don't make it a default right away
- Go the c) route and raise an error

I'm also ok with going the a) route but then we should throw a big warning. 

@LysandreJik @ArthurZucker @amyeroberts wdyt?

Reply:
> Let's add SDPA but don't make it a default right away 

yeah let's do this for now! 
Would rather raise an error as you propose in c) than more warning! Thanks for explaining 😉

Reply:
Let me know your thoughts @LysandreJik @patrickvonplaten @ArthurZucker

Reply:
@patrickvonplaten let me know if the warning sounds good to you